### PR TITLE
LUCENE-8990: Add estimateDocCount(visitor) method to PointValues

### PR DESCRIPTION
**Pull request by iverase**
This PR adds the method estimateDocCount(visitor) to `PointValues` class. This can then be used as a way to estimate the cost() of ScorerSupplier instead of using estimatePointCount() that is actually used.

The reason to that is that other suppliers define it cost in respect of documents match and this can create problems when a numerical field contains more than one value.

### CODE REVIEW

**Comment by jpountz** [comment]
This probably significantly under-estimates the document count for low numbers of points? For instance say that 2 points matched and there are 10 values per document. This formula would return 0.2 while there is more likely 2 documents that matched?

**Comment by iverase** [comment]
Yes the final answer would be 1 because we are rounding up.

I went for underestimate because it is better for IndexAndDocValues query so there is my bias. 

Maybe is a better heuristic to divide the points per document by 2 so it we hit half of the points then we consider one doc. Still in your example the result would be 1. 

Or we should disable indexOrDocValues all together for multi-value fields?

**Comment by jpountz** [comment]
I was more thinking that there should be a better formula, which takes into account the fact that when you match few points, then it's more likely that all points belong to different documents than that they belong to different documents?

**Comment by jpountz** [comment]
@colings86 pointed me to https://math.stackexchange.com/questions/1175295/urn-problem-probability-of-drawing-balls-of-k-unique-colors which seems to have the answer to the problem we are trying to solve, if we can make the assumption that all docs have about the same number of values, which is likely not always the case but still probably a fair assumption for this kind of heuristic?

File: lucene/core/src/test/org/apache/lucene/codecs/lucene60/TestLucene60PointsFormat.java [hunk 1]
```diff
@@ -385,12 +298,102 @@ public Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
...
+  public void testRandomDocCount() {
+    for (int i = 0; i < 100; i++) {
+      long size = TestUtil.nextLong(random(), 1, Long.MAX_VALUE);
+      int maxDoc = (size > Integer.MAX_VALUE) ? Integer.MAX_VALUE : Math.toIntExact(size);
+      int docCount = TestUtil.nextInt(random(), 1, maxDoc);
+      long estimatePointCount = TestUtil.nextLong(random(), 0, size);
+      PointValues values = getPointValues(size, docCount, estimatePointCount);
+      long docs = values.estimateDocCount(null);
+      assertTrue(docs <= estimatePointCount);
+      assertTrue(docs <= maxDoc);
```
Comment (jpountz) [inline lucene/core/src/test/org/apache/lucene/codecs/lucene60/TestLucene60PointsFormat.java]: maybe also assert that `docs >= estimatedPointCount / (size/docCount)`?

**Comment by iverase** [comment]
@jpountz proved me that the solution can be simplified to the following expression if we consider that NumberOfPoints >> NumberValuesPerDoc:

```
D * (1 - ((N - n) / N)^(N/D))
```

where D is the total number of docs, N the total number of values and n the number of counted values.

File: lucene/core/src/java/org/apache/lucene/index/PointValues.java [hunk 2]
```diff
@@ -241,9 +241,28 @@ default void visit(DocIdSetIterator iterator, byte[] packedValue) throws IOExcep
...
+      return docCount;
+    } else if (size == docCount || estimatedPointCount == 0L ) {
+      // if the point count estimate is 0 or we have only single values
+      // return this estimate
+      return  estimatedPointCount;
+    } else {
+      // in case of multi values estimate the number of docs using the solution provided in
+      // https://math.stackexchange.com/questions/1175295/urn-problem-probability-of-drawing-balls-of-k-unique-colors
+      // then approximate the solution for points per doc << size() which results in the expression
+      // D * (1 - ((N - n) / N)^(N/D))
```
[the remark this answers is not in the document]
Comment (iverase) [reply lucene/core/src/java/org/apache/lucene/index/PointValues.java]: It is clarify in the following line.

**Pull request merged**